tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1461 @@
+          promise.SetResult(r);
+          return;
+        }
+      default:
+        break;
+    }
+
+    DoPut();
+  }
+
+  void DoPut() {
+    // TODO(sjperkins). Introduce S3 requester_pays logic here
+    // Assume that if the user_project field is set, that we want to provide
+    // it on the uri for a requestor pays bucket.
+    // AddUserProjectParam(&upload_url, true, owner->encoded_user_project());
+
+    // TODO(sjperkins).
+    // This was changed from POST to PUT as a basic POST does not work
+    // Some more headers need to be added to allow POST to work:
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-authentication-HTTPPOST.html
+    upload_url_ = tensorstore::StrCat(owner->endpoint_, "/", encoded_object_name);
+
+    start_time_ = absl::Now();
+    auto [content_sha256, checksum_sha256] = payload_sha256(value);
+    auto [content_md5, checksum_md5] = payload_md5(value);
+    expected_etag_ = checksum_md5;
+
+    auto request = S3RequestBuilder("PUT", upload_url_)
+            .AddHeader("Content-Type: application/octet-stream")
+            .AddHeader(tensorstore::StrCat("Content-Length: ", value.size()))
+            .AddHeader(absl::StrCat("host: ", owner->host_))
+            .AddHeader(absl::StrCat("x-amz-content-sha256: ", content_sha256))
+            .AddHeader(absl::StrCat("x-amz-checksum-sha256: ", checksum_sha256))
+            .AddHeader(absl::StrCat("x-amz-meta-tensorstore-id: ", content_sha256))
+            .AddHeader(absl::FormatTime("x-amz-date: %Y%m%dT%H%M%SZ", start_time_, absl::UTCTimeZone()))
+            .AddHeader(absl::StrCat("Content-MD5: ", content_md5))
+            .BuildRequest(
+              credentials_.GetAccessKey(),
+              credentials_.GetSecretKey(),
+              owner->aws_region_,
+              content_sha256,
+              start_time_);
+
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_REQUESTS)
+        << "WriteTask: " << request << " size=" << value.size();
+
+    auto future = owner->transport_->IssueRequest(request, value);
+    future.ExecuteWhenReady([self = IntrusivePtr<WriteTask>(this)](
+                                ReadyFuture<HttpResponse> response) {
+      self->OnResponse(response.result());
+    });
+  }
+
+  void OnResponse(const Result<HttpResponse>& response) {
+    if (!promise.result_needed()) {
+      return;
+    }
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_RESPONSES && response.ok())
+        << "WriteTask " << *response;
+
+    absl::Status status = !response.ok() ? response.status()
+                                         : HttpResponseCodeToStatus(response.value());
+
+    if (!status.ok() && IsRetriable(status)) {
+      status = owner->BackoffForAttemptAsync(status, attempt_++, this);
+      if (status.ok()) {
+        return;
+      }
+    }
+    if (!status.ok()) {
+      promise.SetResult(status);
+      return;
+    }
+
+    auto request = S3RequestBuilder("HEAD", upload_url_)
```

Comment:
I put this HEAD request in because, while a PUT returns the correct `ETag`, it doesn't return a `Last-Modified`, only a `Date` which  I don't believe is the same thing. The HEAD response contains both `Last-Modified` and `Etag` which is used to construct a StorageGeneration.